modem_cellular: add U-Blox LARA-R6 and implement change of baudrate

emillindq opened the pull request:
Some modems, including LARA-R6 and Quectel EG-25, have a default baudrate of 115200 despite the modems being multiple MBit/s. This PR adds an optional state in modem_cellular.c which will be entered if the uart-target-speed is set in the DTS, changing the speed and reconfiguring the UART peripheral before continuing on with the init script.

@bjarki-andreasen This is a draft, error checks and such has to be added, but I'd like your input first before I spend more time on this. Please note however that the git diff here on Github is quite confusing...

Thanks!

emillindq committed:
modem_cellular: add U-Blox LARA-R6 and implement change of baudrate

U-Blox LARA-R6 was added to modem_cellular, and an additional state
was introduced where the UART baudrate is changed if the modem supports
it

Signed-off-by: Emil Lindqvist <emil@lindq.gr>

bjarki-andreasen commented:
The target baudrate should not be in the devicetree, it should be a Kconfig. Additionally, the commands to send to set the baudrate are constant, like the APN in the init script. It is quite a bit simpler to add an additonal optional script alongside `const struct modem_chat_script *periodic_chat_script;` in the config, which if not NULL, is run.

emillindq commented:
> . It is quite a bit simpler to add an additonal optional script alongside const struct modem_chat_script *periodic_chat_script; in the config, which if not NULL, is run

I have thought about this a lot, and it is indeed the easy solution. I felt however that hardware-specific configurations should come from the DTS, like it's done in the esp32_at driver
![image](https://github.com/user-attachments/assets/688beee1-38ad-44d8-85c0-f7d730dc1b42)

Anyhow, using Kconfig symbols I was thinking there should be an enable Kconfig, which enables the actual baud rate symbols. But then you'd have to ifdef everywhere in modem_cellular.c wherever the enabled symbols are used. Which is [not so nice](https://github.com/obgm/libcoap/blob/develop/src/coap_io.c).

Anyways, take a look at the latest commit

bjarki-andreasen commented on drivers/modem/modem_cellular.c:
```diff
@@ -717,8 +726,67 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
+		modem_chat_run_script_async(&data->chat, config->set_baudrate_chat_script);
+		break;
+
+	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
+		/* Let modem reconfigure */
+		k_msleep(CONFIG_MODEM_CELLULAR_NEW_BAUDRATE_DELAY);
+		__fallthrough;
```
Delays er performed using modem_cellular_start_timer() and waiting for the `MODEM_CELLULAR_EVENT_TIMEOUT`, sleeping is not allowed in this driver as it blocks the system work queue

emillindq replied:
Makes sense

bjarki-andreasen commented on drivers/modem/modem_cellular.c:
```diff
@@ -1899,6 +1971,70 @@ MODEM_CHAT_SCRIPT_DEFINE(u_blox_sara_r5_periodic_chat_script,
+			 abort_matches, modem_cellular_chat_callback_handler, 1);
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(u_blox_lara_r6_init_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
+			      MODEM_CHAT_SCRIPT_CMD_RESP_NONE("AT", 100),
```
No longer needed? if the modem supports auto baudrate detection (that's what the 4 "AT" are for), you don't need to manually change the baudrate :) If auto baud is not supported, remove the redundant "AT"s

emillindq replied:
True that!

emillindq committed:
modem_cellular: adjust CMUX max frame size for LARA-R6

It has been observed that using a CMUX max frame size of 127
causes FCS errors; smaller or larger works. This commit sets the max
frame size to the modem's default value.

Signed-off-by: Emil Lindqvist <emil@lindq.gr>

emillindq commented:
I threw another small commit into this PR, a CMUX frame size of 127 causes FCS errors. Different checksums even if address, control and length is the same. Must be a bug in the modem, because anything just above or just below doesn't cause FCS errors

bjarki-andreasen commented on drivers/modem/modem_cellular.c:
```diff
@@ -762,10 +830,6 @@ static void modem_cellular_run_init_script_event_handler(struct modem_cellular_d
 		modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_CONNECT_CMUX);
 		break;
 
-	case MODEM_CELLULAR_EVENT_SUSPEND:
-		modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_IDLE);
-		break;
-
```
is this removed intentionally?

rerickson1 commented on drivers/modem/Kconfig.cellular:
```diff
@@ -55,4 +57,16 @@ config MODEM_CELLULAR_USER_PIPE_BUFFER_SIZES
 
+config MODEM_CELLULAR_NEW_BAUDRATE
+	int "New baudrate to configure modem to, if supported"
+	range 0 4000000
```
a range of 0 doesn't make sense does it? I recommend setting the min to something like 9600.

emillindq replied:
fixed

bjarki-andreasen reviewed (dismissed):
I think this looks good :)

rerickson1 commented on drivers/modem/modem_cellular.c:
```diff
@@ -1899,6 +1975,65 @@ MODEM_CHAT_SCRIPT_DEFINE(u_blox_sara_r5_periodic_chat_script,
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CMUX=0,0,5,31", ok_match));
+
+MODEM_CHAT_SCRIPT_DEFINE(u_blox_lara_r6_init_chat_script, u_blox_lara_r6_init_chat_script_cmds,
+			 abort_matches, modem_cellular_chat_callback_handler, 10);
+
```
please check formatting here

emillindq commented on drivers/modem/Kconfig.cellular:
```diff
@@ -55,4 +57,16 @@ config MODEM_CELLULAR_USER_PIPE_BUFFER_SIZES
 	int "The size of the buffers used for each user pipe in bytes."
 	default 128
 
+config MODEM_CELLULAR_NEW_BAUDRATE
+	int "New baudrate to configure modem to, if supported"
+	range 9600 4000000
+	default 3000000 if DT_HAS_U_BLOX_LARA_R6_ENABLED
+	default 115200
```
I guess it depends on both the support of the modem, HW design and UART driver. My vision was that the default would be something that probably works everywhere, and if the user desires something else they can change this in their board's Kconfig.defconfig

rerickson1 replied:
Never mind, `MODEM_BACKEND_UART` already depends on the UART driver, so no need for a depends here.

tomi-font commented on drivers/modem/Kconfig.cellular:
```diff
@@ -55,4 +57,16 @@ config MODEM_CELLULAR_USER_PIPE_BUFFER_SIZES
+	default 115200
+
+config MODEM_CELLULAR_NEW_BAUDRATE_DELAY
+	int "Time modem takes to change baudrate"
```
```suggestion
	int "Time modem takes to change baudrate in milliseconds"
```

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -717,8 +726,67 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
+		 * modem is already at the new baudrate, meaning no reply. So
+		 * ignore any failures and continue as if baudrate is already set
+		 */
+		__fallthrough;
```
Suggest adding a `LOG_DBG()` in this `case`.

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -717,8 +726,67 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
+		cfg.baudrate = CONFIG_MODEM_CELLULAR_NEW_BAUDRATE;
+		uart_configure(config->uart, &cfg);
+
+		modem_chat_release(&data->chat);
+		modem_pipe_attach(data->uart_pipe, modem_cellular_bus_pipe_handler, data);
+		modem_pipe_close_async(data->uart_pipe);
```
Would it make more sense to do the releasing/closing before changing the UART configuration? (not sure)

emillindq replied:
Seems to work, shouldn't really make any practical difference

tomi-font replied:
Curious to know @bjarki-andreasen's viewpoint on this?

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -1899,6 +1975,65 @@ MODEM_CHAT_SCRIPT_DEFINE(u_blox_sara_r5_periodic_chat_script,
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(u_blox_lara_r6_set_baudrate_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP("ATE0", ok_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+IPR="
+					STRINGIFY(CONFIG_MODEM_CELLULAR_NEW_BAUDRATE), ok_match),);
```
minor
```suggestion
					STRINGIFY(CONFIG_MODEM_CELLULAR_NEW_BAUDRATE), ok_match));
```

tomi-font reviewed (changes requested):
Maybe a release notes entry is in order?

tomi-font commented on doc/releases/release-notes-4.0.rst:
```diff
@@ -146,6 +146,9 @@ Drivers and Sensors
 
 * Modem
 
+  * Added support for U-Blox LARA-R6 modem.
```
```suggestion
  * Added support for the U-Blox LARA-R6 modem.
```

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -717,8 +726,77 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
+
+		ret = uart_config_get(config->uart, &cfg);
+		if (ret < 0) {
+			LOG_ERR("Failed to get UART configuration");
+			break;
+		}
+		cfg.baudrate = CONFIG_MODEM_CELLULAR_NEW_BAUDRATE;
+		ret = uart_configure(config->uart, &cfg);
+		if (ret < 0) {
+			LOG_ERR("Failed to set new baudrate");
```
Those logs could log the error code to be even more helpful.

emillindq replied:
Why not 🙂
